### PR TITLE
allow setting of dirMode and fileMode when mounting azure file volume in pod

### LINKED ISSUE #37005: azure_file volumes should allow setting of dir_mode and file_mode
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.): no

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): azurefile azure_file 0777

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): FEATURE REQUEST

**Kubernetes version** (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.5", GitCommit:"5a0a696437ad35c133c0c8493f7e9d22b0f9b81b", GitTreeState:"clean", BuildDate:"2016-10-29T01:38:40Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.5", GitCommit:"5a0a696437ad35c133c0c8493f7e9d22b0f9b81b", GitTreeState:"clean", BuildDate:"2016-10-29T01:32:42Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}

**Environment**:
- **Cloud provider or hardware configuration**: Microsoft Azure Container Service
- **OS** (e.g. from /etc/os-release): Ubuntu 16.04 LTS
- **Kernel** (e.g. `uname -a`): Linux 4.4.0-28-generic #47-Ubuntu SMP Fri Jun 24 10:09:13 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

**What happened**:
I attempted to set up an Azure File Volume using the built-in volume type azureFile.  Everything works, except that the files and directories defaults to permission 0777, always.

When I tried to manually mount within the container, I could customise dir_mode and file_mode, meaning that this change should be possible.  Reading the source code confirms it - dir_mode=0777 and file_mode=0777 seem hardcoded.

**What you expected to happen**:
Being able to supply dir_mode and file_mode to the mount. e.g.
```
- name: myvolume
  azureFile:
    secretName: supersecret
    shareName: myshare
    readOnly: false
    dirMode: 0400
    fileMode: 0400
```
**How to reproduce it** (as minimally and precisely as possible):
N/A.  This is a feature request.

**Anything else do we need to know**:
N/A

### PR DESCRIPTION
**Pull request by andyzhangx**
**Which issue this PR fixes**:
Fixes [#37005: azure_file volumes should allow setting of dir_mode and file_mode](https://github.com/kubernetes/kubernetes/issues/37005)

**What this PR does / why we need it**:
This PR is to allow setting of dirMode and fileMode permission when mounting azure file volume in pod. The permission is hard coded to 0777 without this PR.

**Special notes for your reviewer**:
This PR will add two more fields(dirMode, fileMode) in azure_file/azure.yaml, as json file only supports decimal, 0777 is the default value.
- *dirMode*: Directory permission code that gives rwx permissions to all users. Must be a value between 0 and 0777. Defaults to 0777.
- *fileMode*: File permission code that gives rwx permissions to all users. Must be a value between 0 and 0777. Defaults to 0777.

```
  name: myvolume
  azureFile:
    secretName: supersecret
    shareName: myshare
    readOnly: false
    dirMode: 0777
    fileMode: 0777
```
**Release note**:

### CODE REVIEW

**Comment by spiffxp** [comment]
@k8s-bot ok to test

File: staging/src/k8s.io/client-go/pkg/api/types.go [hunk 1]
```diff
@@ -1009,6 +1009,12 @@ type AzureFileVolumeSource struct {
 	// the ReadOnly setting in VolumeMounts.
 	// +optional
 	ReadOnly bool
+	// Directory permission code that gives rwx permissions to all users. Must be a value between 0 and 511(0777). Defaults to 511(0777).
+	// +optional
+	DirMode int32
```
Comment (timstclair) [inline staging/src/k8s.io/client-go/pkg/api/types.go]: Type should be `*int32`, same below.

File: staging/src/k8s.io/client-go/pkg/api/types.go [hunk 2]
```diff
@@ -1009,6 +1009,12 @@ type AzureFileVolumeSource struct {
 	// the ReadOnly setting in VolumeMounts.
 	// +optional
 	ReadOnly bool
+	// Directory permission code that gives rwx permissions to all users. Must be a value between 0 and 511(0777). Defaults to 511(0777).
```
Comment (timstclair) [inline staging/src/k8s.io/client-go/pkg/api/types.go]: Valid range should be enforced in [validation.go](https://github.com/kubernetes/kubernetes/blob/448f8501596ef0c14e2720583385d5172f209105/pkg/api/validation/validation.go#L997) and default should be set in [defaults.go](https://github.com/kubernetes/kubernetes/blob/448f8501596ef0c14e2720583385d5172f209105/pkg/api/v1/defaults.go#L229)

**Review by timstclair — CHANGES_REQUESTED** [review]
Please review https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md, it lists all the changes you need to make.

@kubernetes/sig-storage-api-reviews @kubernetes/api-approvers

File: staging/src/k8s.io/client-go/pkg/api/types.go (hunk 2 above)
Comment (andyzhangx) [inline staging/src/k8s.io/client-go/pkg/api/types.go]: fixed

File: staging/src/k8s.io/client-go/pkg/api/types.go (hunk 1 above)
Comment (andyzhangx) [inline staging/src/k8s.io/client-go/pkg/api/types.go]: fixed

File: staging/src/k8s.io/client-go/pkg/api/types.go (hunk 2 above)
Comment (andyzhangx) [inline staging/src/k8s.io/client-go/pkg/api/types.go]: I have done the validation in validation.go and set default in defaults.go

**Comment by andyzhangx** [comment]
@timstclair Thanks for your instructions, I am a newbie for k8s, just took me a lot of efforts to fix all this, while I think it's only a very tiny fix in the begging. Would you take a review of all this? thanks.

**Comment by rootfs** [comment]
dirMod and fileMod are among the end user visible options, there are still others users want to change. Making api changes for mount options is not scalable.

[Mount option](https://github.com/kubernetes/features/issues/168) support makes it possible to set these options without changing api object

**Pull request closed without merging**